[AppKit] Add some of the missing NSAccessibility bits

iainx opened the pull request.

dnfclas commented:
Hi __@iainx__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla2.dotnetfoundation.org.
    </span>

TTYL, DNFBOT;

spouliot commented on src/AppKit/NSAccessibility.cs:
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
+		[DllImport (Constants.AppKitLibrary)]
+		static extern IntPtr NSAccessibilityRoleDescription (IntPtr role, IntPtr subrole);
+
+		public static string RoleDescription (NSString role, NSString subrole)
```
a method name needs a verb, e.g. `GetRoleDescription`

spouliot commented on src/AppKit/NSAccessibility.cs:
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
+
+		public static void PostNotification (NSObject element, NSString notification, NSDictionary userInfo)
+		{
+			NSAccessibilityPostNotificationWithUserInfo (element.Handle, notification.Handle, userInfo.Handle);
```
Any `null` arguments will cause NRE (common issue in several API here).

If `null` is allowed by the native API then `IntPtr.Zero` must be given to p/invoke.
If not then a null check with a ANE should be thrown.

spouliot commented on src/AppKit/NSAccessibility.cs:
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
+			IntPtr handle = NSAccessibilityRoleDescriptionForUIElement (element.Handle);
+
+			if (handle == IntPtr.Zero)
+				return null;
```
that check is not needed, `NSString.FromHandle` already _handle_ it

spouliot commented:
@iainx thanks for the PR :) my 2 comments (method name and null check) needs to be applied globally. 

Also can you confirm that you ran the introspection tests with those changes ? that will spot some potential defects (including incorrect Apple documentation). If you did just reply here, thanks!

Also @chamons might have more feedback too...

xamarin-release-manager commented:
Build success

chamons commented on src/AppKit/NSAccessibility.cs:
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
+		[DllImport (Constants.AppKitLibrary)]
+		static extern IntPtr NSAccessibilityActionDescription (IntPtr action);
+
+		public static string ActionDescription (NSString action)
```
Another "needs a verb" in name

chamons commented on src/AppKit/NSAccessibility.cs:
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
+		[DllImport (Constants.AppKitLibrary)]
+		static extern IntPtr NSAccessibilityUnignoredAncestor (IntPtr element);
+
+		public static NSObject UnignoredAncestor (NSObject element)
```
Again.

chamons commented on src/AppKit/NSAccessibility.cs:
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
+		[DllImport (Constants.AppKitLibrary)]
+		static extern IntPtr NSAccessibilityUnignoredDescendant (IntPtr element);
+
+		public static NSObject UnignoredDescendant (NSObject element)
```
Again.

chamons commented on src/AppKit/NSAccessibility.cs:
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
+		[DllImport (Constants.AppKitLibrary)]
+		static extern IntPtr NSAccessibilityUnignoredChildren (IntPtr originalChildren);
+
+		public static NSObject[] UnignoredChildren (NSArray originalChildren)
```
Again.

chamons commented on src/AppKit/NSAccessibility.cs:
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
+		[DllImport (Constants.AppKitLibrary)]
+		static extern IntPtr NSAccessibilityUnignoredChildrenForOnlyChild (IntPtr originalChild);
+
+		public static NSObject[] UnignoredChildrenForOnlyChild (NSObject originalChild)
```
Again.

chamons commented:
I picked out the places where @spouliot's name change needs hit. Once you land those then LGTM.

spouliot commented on src/appkit.cs:
```diff
@@ -21879,4 +22792,39 @@ public interface NSAccessibilityLayoutItem : NSAccessibilityGroup {
+	[Category, BaseType (typeof (NSObject))]
+	public interface NSObjectAccessibilityExtensions {
+		[Export ("accessibilityHitTest:")]
+		NSObject AccessibilityHitTest (CGPoint point);
```
same for binding categories, methods needs verbs :)

rolfbjarne commented on src/AppKit/NSAccessibility.cs:
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
+		public static NSObject UnignoredAncestor (NSObject element)
+		{
+			var handle = NSAccessibilityUnignoredAncestor (element.Handle);
+			return new NSObject (handle);
```
Use `Runtime.GetNSObject (handle)` instead, which does not create a new managed wrapper if one already exists.

rolfbjarne commented on src/AppKit/NSAccessibility.cs:
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
+		{
+			var handle = NSAccessibilityUnignoredDescendant (element.Handle);
+
+			return new NSObject (handle);
```
Use `Runtime.GetNSObject (handle)` instead.

mandel-macaque commented on src/AppKit/NSAccessibility.cs:
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
+		public static void PostNotification (NSObject element, NSString notification)
+		{
+			NSAccessibilityPostNotification (element.Handle, notification.Handle);
+		}
```
I believe you need to check against null ptr here too. element and notification can be null and you will get a NRE.

mandel-macaque commented on src/AppKit/NSAccessibility.cs:
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
+		public static string RoleDescription (NSObject element)
+		{
+			IntPtr handle = NSAccessibilityRoleDescriptionForUIElement (element.Handle);
+
```
A check is indeed needed for element AFAIK, else null.Handler will throw an exception.

mandel-macaque commented on src/appkit.cs:
```diff
@@ -21624,6 +21624,919 @@ partial interface NSAccessibilityAttributes {
+
+	[Static]
+	public interface NSAccessibilityFontKeys {
+		[Field ("NSAccessibilityFontNameKey")]
```
What is the usage of this keys? Are they used to access data in an NSDictionary? Is it is the case we should use a strong dictionary.

iainx replied:
Yes, the FontKeys are used with the NSDictionary returned using the FontTextAttribute.

iainx replied:
But the NSDictionary is returned by the (obsoleted) GetAccessibilityAttributeValue method, which returns a generic NSObject, so we can't using the strong dictionary here either I'm afraid.

mandel-macaque commented on src/appkit.cs:
```diff
@@ -21624,6 +21624,919 @@ partial interface NSAccessibilityAttributes {
+
+	[Static]
+	public interface NSAccessibilityNotificationUserInfoKeys {
+		[Field ("NSAccessibilityUIElementsKey")]
```
Same question about the usage of the keys.

iainx replied:
Again, yes, these are keys used in the NSDictionary user info passed to the PostNotification method

iainx replied:
But I think other custom keys can be used in the UserInfo dictionary too, so making a strong dictionary wouldn't be possible.

iainx commented:
I've fixed all the bits mentioned and I've run the introspection tests.

I changed the NSObjectAccessibilityExtensions and NSWorkspaceAccessibilityExtensions to be a protocol instead of a category because it's really an informal protocol implemented as a category.

I also added all the obsolete methods because in standard Apple style they've obsoleted things without giving a better way of doing that task, so it looks like the obsolete methods will still be needed for certain things.

xamarin-release-manager commented:
Build success

rolfbjarne commented:
My comments have been solved to my satisfaction, and I have nothing else 👍

spouliot commented on src/appkit.cs:
```diff
@@ -21879,4 +22792,79 @@ public interface NSAccessibilityLayoutItem : NSAccessibilityGroup {
+
+	[Protocol, BaseType (typeof (NSObject))]
+	public interface NSObjectAccessibilityExtensions {
+		[Obsolete]
```
That's not really helpful for developers as it does not say why ? and what's the recommended replacement ?

In most cases that's better handled with something like:

`[Availability (Obsoleted = Platform.Mac_10_10, Message = "Use XXX instead")]`

as it tells when it was obsoleted and how to update the code (when possible). IIRC you mentioned some things were obsolete without an alternative and, if that's the case, then it should become the text, e.g. "No alternative is documented".

Also in some cases it's `Deprecated` and not `Obsoleted` that should be used (or both).

iainx commented:
I've fixed the Obsolete attributes, directing people at the newer protocols, which is all Apple's docs do too, as there's no direct equivalent (it seems they're wanting to do things in a different way from 10.10)

I'm not sure what the difference between Obsoleted and Deprecated is so maybe I should have used Deprecated.

xamarin-release-manager commented:
Build success

chamons commented:
I agree with @rolfbjarne that everything looks reasonable now. Landing.

spouliot commented:
@chamons dontlink-mac fails since this was merged https://wrench.internalx.com/Wrench/ViewLane.aspx?lane_id=1381&host_id=243&revision_id=768118

iainx commented:
@spouliot @chamons yes, sorry the accessibilityAttributeNames shouldn't have the :
I'll make a new PR to fix that tomorrow

iainx commented:
Is it possible to run these tests locally?

chamons commented:
@iainx see https://github.com/xamarin/xamarin-macios/pull/143

and make run-mac-dontlink in tests dir
